In [21]:
import mysql.connector as conn
import pandas as pd

In [30]:
myconn = conn.connect(host = "localhost", user = "root", passwd = "12345")

In [118]:
myconn

In [31]:
cursor = myconn.cursor()

In [5]:
cursor

In [91]:
result = cursor.execute('show databases')

In [92]:
cursor.fetchall()

[('assignment',),
 ('information_schema',),
 ('mysql',),
 ('performance_schema',),
 ('sakila',),
 ('sys',),
 ('world',)]

In [93]:
query = 'use world;'

cursor.execute(query)

In [94]:
result = 'show tables;'

cursor.execute(result)

print(cursor.fetchall())

[('city',), ('country',), ('countrylanguage',)]


In [95]:
query1 = """
SELECT * 
FROM city
"""

data = pd.read_sql_query(query1, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\1017593583.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query1, myconn)


ID            Name CountryCode       District  Population
0        1           Kabul         AFG          Kabol     1780000
1        2        Qandahar         AFG       Qandahar      237500
2        3           Herat         AFG          Herat      186800
3        4  Mazar-e-Sharif         AFG          Balkh      127800
4        5       Amsterdam         NLD  Noord-Holland      731200
...    ...             ...         ...            ...         ...
4074  4075      Khan Yunis         PSE     Khan Yunis      123175
4075  4076          Hebron         PSE         Hebron      119401
4076  4077        Jabaliya         PSE     North Gaza      113901
4077  4078          Nablus         PSE         Nablus      100231
4078  4079           Rafah         PSE          Rafah       92020

[4079 rows x 5 columns]

In [96]:
query2 = """
SELECT * 
FROM country
"""

data = pd.read_sql_query(query2, myconn)

print(data.shape)
data.head(2)

(239, 15)


C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\3280154561.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query2, myconn)


Code         Name      Continent                     Region  SurfaceArea  \
0  ABW        Aruba  North America                  Caribbean        193.0   
1  AFG  Afghanistan           Asia  Southern and Central Asia     652090.0   

   IndepYear  Population  LifeExpectancy     GNP  GNPOld  \
0        NaN      103000            78.4   828.0   793.0   
1     1919.0    22720000            45.9  5976.0     NaN   

               LocalName                                GovernmentForm  \
0                  Aruba  Nonmetropolitan Territory of The Netherlands   
1  Afganistan/Afqanestan                               Islamic Emirate   

     HeadOfState  Capital Code2  
0        Beatrix    129.0    AW  
1  Mohammad Omar      1.0    AF

In [97]:
data.to_csv('city.csv', index = False)

In [98]:
data[data['Continent'] == 'Asia']

Code                  Name Continent                     Region  \
1    AFG           Afghanistan      Asia  Southern and Central Asia   
7    ARE  United Arab Emirates      Asia                Middle East   
9    ARM               Armenia      Asia                Middle East   
16   AZE            Azerbaijan      Asia                Middle East   
21   BGD            Bangladesh      Asia  Southern and Central Asia   
23   BHR               Bahrain      Asia                Middle East   
32   BRN                Brunei      Asia             Southeast Asia   
33   BTN                Bhutan      Asia  Southern and Central Asia   
41   CHN                 China      Asia               Eastern Asia   
54   CYP                Cyprus      Asia                Middle East   
77   GEO               Georgia      Asia                Middle East   
92   HKG             Hong Kong      Asia               Eastern Asia   
98   IDN             Indonesia      Asia             Southeast Asia   
99   IND                 India      Asia  Southern and Central Asia   
102  IRN                  Iran      Asia  Southern and Central Asia   
103  IRQ                  Iraq      Asia                Middle East   
105  ISR                Israel      Asia                Middle East   
108  JOR                Jordan      Asia                Middle East   
109  JPN                 Japan      Asia               Eastern Asia   
110  KAZ             Kazakstan      Asia  Southern and Central Asia   
112  KGZ            Kyrgyzstan      Asia  Southern and Central Asia   
113  KHM              Cambodia      Asia             Southeast Asia   
116  KOR           South Korea      Asia               Eastern Asia   
117  KWT                Kuwait      Asia                Middle East   
118  LAO                  Laos      Asia             Southeast Asia   
119  LBN               Lebanon      Asia                Middle East   
124  LKA             Sri Lanka      Asia  Southern and Central Asia   
129  MAC                 Macao      Asia               Eastern Asia   
134  MDV              Maldives      Asia  Southern and Central Asia   
140  MMR               Myanmar      Asia             Southeast Asia   
141  MNG              Mongolia      Asia               Eastern Asia   
149  MYS              Malaysia      Asia             Southeast Asia   
160  NPL                 Nepal      Asia  Southern and Central Asia   
163  OMN                  Oman      Asia                Middle East   
164  PAK              Pakistan      Asia  Southern and Central Asia   
168  PHL           Philippines      Asia             Southeast Asia   
173  PRK           North Korea      Asia               Eastern Asia   
176  PSE             Palestine      Asia                Middle East   
178  QAT                 Qatar      Asia                Middle East   
183  SAU          Saudi Arabia      Asia                Middle East   
186  SGP             Singapore      Asia             Southeast Asia   
203  SYR                 Syria      Asia                Middle East   
207  THA              Thailand      Asia             Southeast Asia   
208  TJK            Tajikistan      Asia  Southern and Central Asia   
210  TKM          Turkmenistan      Asia  Southern and Central Asia   
211  TMP            East Timor      Asia             Southeast Asia   
215  TUR                Turkey      Asia                Middle East   
217  TWN                Taiwan      Asia               Eastern Asia   
224  UZB            Uzbekistan      Asia  Southern and Central Asia   
230  VNM               Vietnam      Asia             Southeast Asia   
234  YEM                 Yemen      Asia                Middle East   

     SurfaceArea  IndepYear  Population  LifeExpectancy        GNP     GNPOld  \
1       652090.0     1919.0    22720000            45.9     5976.0        NaN   
7        83600.0     1971.0     2441000            74.1    37966.0    36846.0   
9        29800.0     1991.0     3520000            66.4     1813.0  

In [99]:
data[data["LocalName"] == 'Aruba']

Code   Name      Continent     Region  SurfaceArea  IndepYear  Population  \
0  ABW  Aruba  North America  Caribbean        193.0        NaN      103000   

   LifeExpectancy    GNP  GNPOld LocalName  \
0            78.4  828.0   793.0     Aruba   

                                 GovernmentForm HeadOfState  Capital Code2  
0  Nonmetropolitan Territory of The Netherlands     Beatrix    129.0    AW

### Read Selected Columns

In [100]:
query = """
SELECT Continent, Population
FROM country
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\1474015279.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


Continent  Population
0    North America      103000
1             Asia    22720000
2           Africa    12878000
3    North America        8000
4           Europe     3401200
..             ...         ...
234           Asia    18112000
235         Europe    10640000
236         Africa    40377000
237         Africa     9169000
238         Africa    11669000

[239 rows x 2 columns]

In [101]:
list(data['Continent'].unique())

['North America',
 'Asia',
 'Africa',
 'Europe',
 'South America',
 'Oceania',
 'Antarctica']

In [102]:
data['Continent'].value_counts()

Continent
Africa           58
Asia             51
Europe           46
North America    37
Oceania          28
South America    14
Antarctica        5
Name: count, dtype: int64

### Extract unique values from column

In [103]:
query = """
SELECT DISTINCT Continent  
FROM country
"""

# unique values query directly

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\2892303321.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


Continent
0  North America
1           Asia
2         Africa
3         Europe
4  South America
5        Oceania
6     Antarctica

In [104]:
data['Continent'].value_counts()

Continent
North America    1
Asia             1
Africa           1
Europe           1
South America    1
Oceania          1
Antarctica       1
Name: count, dtype: int64

In [120]:
query = 'use world;'

cursor.execute(query)

In [121]:
query = 'show tables;'

cursor.execute(query)

print(cursor.fetchall())

[('city',), ('country',), ('countrylanguage',)]


In [122]:
query = """
SELECT * 
FROM city
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\4002667956.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID            Name CountryCode       District  Population
0        1           Kabul         AFG          Kabol     1780000
1        2        Qandahar         AFG       Qandahar      237500
2        3           Herat         AFG          Herat      186800
3        4  Mazar-e-Sharif         AFG          Balkh      127800
4        5       Amsterdam         NLD  Noord-Holland      731200
...    ...             ...         ...            ...         ...
4074  4075      Khan Yunis         PSE     Khan Yunis      123175
4075  4076          Hebron         PSE         Hebron      119401
4076  4077        Jabaliya         PSE     North Gaza      113901
4077  4078          Nablus         PSE         Nablus      100231
4078  4079           Rafah         PSE          Rafah       92020

[4079 rows x 5 columns]

In [126]:
query = """
SELECT *
FROM city
WHERE Population > 123175
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\2349157743.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID            Name CountryCode       District  Population
0        1           Kabul         AFG          Kabol     1780000
1        2        Qandahar         AFG       Qandahar      237500
2        3           Herat         AFG          Herat      186800
3        4  Mazar-e-Sharif         AFG          Balkh      127800
4        5       Amsterdam         NLD  Noord-Holland      731200
...    ...             ...         ...            ...         ...
2835  4070     Chitungwiza         ZWE         Harare      274912
2836  4071    Mount Darwin         ZWE         Harare      164362
2837  4072          Mutare         ZWE     Manicaland      131367
2838  4073           Gweru         ZWE       Midlands      128037
2839  4074            Gaza         PSE           Gaza      353632

[2840 rows x 5 columns]

In [129]:
data['Population'].unique()

array([1780000,  237500,  186800, ...,  131367,  128037,  353632],
      dtype=int64)

In [130]:
data

ID            Name CountryCode       District  Population
0        1           Kabul         AFG          Kabol     1780000
1        2        Qandahar         AFG       Qandahar      237500
2        3           Herat         AFG          Herat      186800
3        4  Mazar-e-Sharif         AFG          Balkh      127800
4        5       Amsterdam         NLD  Noord-Holland      731200
...    ...             ...         ...            ...         ...
2835  4070     Chitungwiza         ZWE         Harare      274912
2836  4071    Mount Darwin         ZWE         Harare      164362
2837  4072          Mutare         ZWE     Manicaland      131367
2838  4073           Gweru         ZWE       Midlands      128037
2839  4074            Gaza         PSE           Gaza      353632

[2840 rows x 5 columns]

### More than one conditions

In [131]:
query = """
SELECT * 
FROM city
WHERE Population > 128037 
AND CountryCode = 'ZWE'
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\104196079.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID          Name CountryCode    District  Population
0  4068        Harare         ZWE      Harare     1410000
1  4069      Bulawayo         ZWE    Bulawayo      621742
2  4070   Chitungwiza         ZWE      Harare      274912
3  4071  Mount Darwin         ZWE      Harare      164362
4  4072        Mutare         ZWE  Manicaland      131367

In [132]:
query = """
SELECT * 
FROM city 
WHERE Population > 131367
OR District = 'Harare'
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\3949183196.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID          Name CountryCode       District  Population
0        1         Kabul         AFG          Kabol     1780000
1        2      Qandahar         AFG       Qandahar      237500
2        3         Herat         AFG          Herat      186800
3        5     Amsterdam         NLD  Noord-Holland      731200
4        6     Rotterdam         NLD   Zuid-Holland      593321
...    ...           ...         ...            ...         ...
2606  4068        Harare         ZWE         Harare     1410000
2607  4069      Bulawayo         ZWE       Bulawayo      621742
2608  4070   Chitungwiza         ZWE         Harare      274912
2609  4071  Mount Darwin         ZWE         Harare      164362
2610  4074          Gaza         PSE           Gaza      353632

[2611 rows x 5 columns]

In [135]:
query = """
SELECT *
FROM city
WHERE CountryCode = 'IND'
AND Name = 'Delhi'
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\4085134944.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID   Name CountryCode District  Population
0  1025  Delhi         IND    Delhi     7206704

In [139]:
my_city_list = ['Delhi', 'Mumbai', 'Cheannai']


query = f"""
SELECT *
FROM city
WHERE CountryCode = 'IND'
AND Name IN {tuple(my_city_list)}
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\1186515591.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID   Name CountryCode District  Population
0  1025  Delhi         IND    Delhi     7206704

### Get result in sorted form

In [141]:
query = """
SELECT * 
FROM city
ORDER BY Population ASC
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\433362295.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID                Name CountryCode      District  Population
0     2912           Adamstown         PCN             –          42
1     2317         West Island         CCK   West Island         167
2     3333             Fakaofo         TKL       Fakaofo         300
3     3538  Città del Vaticano         VAT             –         455
4     2316              Bantam         CCK   Home Island         503
...    ...                 ...         ...           ...         ...
4074   939             Jakarta         IDN  Jakarta Raya     9604900
4075  1890            Shanghai         CHN      Shanghai     9696300
4076   206           São Paulo         BRA     São Paulo     9968485
4077  2331               Seoul         KOR         Seoul     9981619
4078  1024     Mumbai (Bombay)         IND   Maharashtra    10500000

[4079 rows x 5 columns]

### Read data from city with name is ascending and population in descending order

In [143]:
query = """
SELECT *
FROM city
ORDER BY Name ASC, Population DESC
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_20132\1846548437.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID                          Name CountryCode             District  \
0      698  [San Cristóbal de] la Laguna         ESP       Canary Islands   
1       20              ´s-Hertogenbosch         NLD        Noord-Brabant   
2      670          A Coruña (La Coruña)         ESP              Galicia   
3     3097                        Aachen         DEU  Nordrhein-Westfalen   
4     3318                       Aalborg         DNK          Nordjylland   
...    ...                           ...         ...                  ...   
4074  2025                         Zunyi         CHN              Guizhou   
4075  3245                        Zürich         CHE               Zürich   
4076  3145                       Zwickau         DEU                Saksi   
4077    28                        Zwolle         NLD           Overijssel   
4078  3446                       Zytomyr         UKR              Zytomyr   

      Population  
0         127945  
1         129170  
2         243402  
3         243825  
4         161161  
...          ...  
4074      261862  
4075      336800  
4076      104146  
4077      105819  
4078      297000  

[4079 rows x 5 columns]

In [36]:
cursor.execute('USE sakila;')

### NULL statement

In [8]:
query = """
SELECT *
FROM customer
WHERE first_name IS NULL
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\4221157929.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


Empty DataFrame
Columns: [customer_id, store_id, first_name, last_name, email, address_id, active, create_date, last_update]
Index: []

### NOT NULL Statement

In [12]:
query = """
SELECT *
FROM customer
WHERE first_name IS NOT NULL
"""

data = pd.read_sql_query(query, myconn)

data.head(2)

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\3680727661.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


customer_id  store_id first_name last_name  \
0            1         1      AMOGH    PATHAK   
1            2         1   PATRICIA   JOHNSON   

                                 email  address_id  active  \
0                  amogh9792@gmail.com           5       1   
1  PATRICIA.JOHNSON@sakilacustomer.org           6       1   

          create_date         last_update  
0 2006-02-14 22:04:36 2023-12-16 19:29:31  
1 2006-02-14 22:04:36 2006-02-15 04:57:20

### UPDATE Values

In [13]:
query = """
UPDATE customer
SET first_name = 'VIJAY', last_name = 'PATHAK'
WHERE customer_id = 1
"""

cursor.execute(query)

In [14]:
query = """
SELECT * 
FROM customer 
WHERE customer_id = 1
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\960006337.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


customer_id  store_id first_name last_name                email  \
0            1         1      VIJAY    PATHAK  amogh9792@gmail.com   

   address_id  active         create_date         last_update  
0           5       1 2006-02-14 22:04:36 2023-12-23 14:43:56

### Create table

In [68]:
query = """
CREATE TABLE TABLE8(
    attendance_number INT,
    first_name VARCHAR(50),
    last_name VARCHAR(50),
    percentage DECIMAL(5,2),
    std INT, 
    division VARCHAR(20)
)
"""

cursor.execute(query)

In [69]:
result = cursor.execute('show tables')

cursor.fetchall() 

[('actor',),
 ('actor_info',),
 ('address',),
 ('category',),
 ('city',),
 ('country',),
 ('customer',),
 ('customer_list',),
 ('film',),
 ('film_actor',),
 ('film_category',),
 ('film_list',),
 ('film_text',),
 ('inventory',),
 ('language',),
 ('nicer_but_slower_film_list',),
 ('payment',),
 ('rental',),
 ('sales_by_film_category',),
 ('sales_by_store',),
 ('staff',),
 ('staff_list',),
 ('store',),
 ('student_datawave',),
 ('table8',)]

### Insert records

In [37]:
query = """
INSERT INTO student_datawave(attendance_number,first_name,last_name,percentage,std,division)
VALUES(1, "amogh","pathak",98.5, 12, 'A')
"""

cursor.execute(query)

In [38]:
query = """
SELECT * 
FROM student_datawave
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\3458914233.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


attendance_number first_name last_name  percentage  std division
0                  1      amogh    pathak        98.5   12        A

### Multiple record insertion

In [46]:
query = """
INSERT INTO student_datawave(attendance_number,first_name,last_name,percentage,std,division)
VALUES(2, 'amol', 'Sharma', 89, 73, 'B'),
      (3, 'vaibhav', 'patil', 99,2, 'Z'),
      (4, 'aaditya', 'Singh', 98, 0, 'Z')
"""

cursor.execute(query)

ProgrammingError: 1146 (42S02): Table 'assign.student_datawave' doesn't exist

In [44]:
query = """
SELECT *
FROM student_datawave
"""

data = pd.read_sql_query(query, myconn)
data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\571638982.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


attendance_number first_name last_name  percentage  std division
0                  1      amogh    pathak        98.5   12        A
1                  2       amol    Sharma        89.0   73        B
2                  3    vaibhav     patil        99.0    2        Z
3                  4    aaditya     Singh        98.0    0        Z

### DELETE

In [45]:
query = """
DELETE FROM student_datawave
WHERE last_name = "Sharma"
"""

cursor.execute(query)

In [46]:
query = """
SELECT *
FROM student_datawave
"""

data = pd.read_sql_query(query, myconn)
data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\571638982.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


attendance_number first_name last_name  percentage  std division
0                  1      amogh    pathak        98.5   12        A
1                  3    vaibhav     patil        99.0    2        Z
2                  4    aaditya     Singh        98.0    0        Z

### Drop Table

In [51]:
query = """
DROP TABLE IF EXISTS TABLE1
"""

cursor.execute(query)

In [71]:
result = cursor.execute('show tables')

cursor.fetchall() 

[('actor',),
 ('actor_info',),
 ('address',),
 ('category',),
 ('city',),
 ('country',),
 ('customer',),
 ('customer_list',),
 ('film',),
 ('film_actor',),
 ('film_category',),
 ('film_list',),
 ('film_text',),
 ('inventory',),
 ('language',),
 ('nicer_but_slower_film_list',),
 ('payment',),
 ('rental',),
 ('sales_by_film_category',),
 ('sales_by_store',),
 ('staff',),
 ('staff_list',),
 ('store',),
 ('student_datawave',)]

### Drop multiple tables

In [70]:
table_to_drop = ['table4', 'table5','table6', 'table7', 'table8']

for t in table_to_drop:
    query = f"""
    DROP TABLE IF EXISTS {t}
    """
    cursor.execute(query)

### LIMIT

In [73]:
cursor.execute('USE sakila')

query = """
SELECT * 
FROM student_datawave
LIMIT 1
"""

data = pd.read_sql_query(query, myconn)

print(f"shape: {data.shape}")

data

shape: (1, 6)


C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\1403528560.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


attendance_number first_name last_name  percentage  std division
0                  1      amogh    pathak        98.5   12        A

### AGGREGATION

MIN
MAX 
AVG 
COUNT
Sum

In [75]:
cursor.execute('USE world')

query = """
SELECT MAX(Population)
FROM city
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\903437614.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


MAX(Population)
0         10500000

In [76]:
cursor.execute('USE world')

query = """
SELECT MIN(Population)
FROM city
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\2025904714.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


MIN(Population)
0               42

In [77]:
cursor.execute('USE world')

query = """
SELECT count(*)  
FROM city
"""

# COUNT

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\2666032921.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


count(*)
0      4079

In [78]:
cursor.execute('USE world')

query = """
SELECT count(*)
FROM city
WHERE Population < 1000000
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\241661348.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


count(*)
0      3841

### AVERAGE

In [79]:
cursor.execute('USE world')

query = """
SELECT AVG(Population)
FROM city
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\103760809.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


AVG(Population)
0      350468.2236

### SUM

In [80]:
cursor.execute('USE world')

query = """
SELECT SUM(Population)
FROM city
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\2989920353.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


SUM(Population)
0     1.429560e+09

### LIKE OPERATOR

### % : zero or one or multiple characters
### _ : single character

In [82]:
query = """
SELECT * 
FROM city
WHERE Name LIKE 'M%'
"""

data = pd.read_sql_query(query, myconn)

print(f"shape: {data.shape}")

data.head(10)

shape: (301, 5)


C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\4007328170.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID                 Name CountryCode      District  Population
0    4       Mazar-e-Sharif         AFG         Balkh      127800
1   22           Maastricht         NLD       Limburg      122087
2   46           Mostaganem         DZA    Mostaganem      115212
3   77        Mar del Plata         ARG  Buenos Aires      512880
4   80                Merlo         ARG  Buenos Aires      463846
5   83               Moreno         ARG  Buenos Aires      356993
6   87                Morón         ARG  Buenos Aires      349246
7   91  Malvinas Argentinas         ARG  Buenos Aires      290335
8  116              Mendoza         ARG       Mendoza      123027
9  131            Melbourne         AUS      Victoria     2865329

In [ ]:
# name starts with M

query = """
SELECT * 
FROM city
WHERE Name LIKE 'M%' 
"""

data = pd.read_sql_query(query, myconn)

print(f"shape: {data.shape}")

data.head(10)

In [ ]:
# name starts with second character is M

In [83]:
query = """
SELECT * 
FROM city
WHERE Name LIKE '_M%' 
"""

data = pd.read_sql_query(query, myconn)

print(f"shape: {data.shape}")

data.head(10)

shape: (36, 5)


C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\841334963.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID        Name CountryCode          District  Population
0    5   Amsterdam         NLD     Noord-Holland      731200
1   21  Amersfoort         NLD           Utrecht      126270
2   27       Emmen         NLD           Drenthe      105853
3  299  Imperatriz         BRA          Maranhão      224564
4  301        Embu         BRA         São Paulo      222223
5  326   Americana         BRA         São Paulo      177409
6  599      Ambato         ECU        Tungurahua      169612
7  726      Umlazi         ZAF     KwaZulu-Natal      339233
8  809        Imus         PHL  Southern Tagalog      195482
9  971       Ambon         IDN           Molukit      249312

### IN OPERATOR

In [84]:
query = """
SELECT *
FROM city
WHERE Name IN ('Delhi', 'Pune', 'Mumbai')
"""

data = pd.read_sql_query(query, myconn)

print(f"shape: {data.shape}")

data.head(10)

shape: (2, 5)


C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\2500756921.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


ID   Name CountryCode     District  Population
0  1025  Delhi         IND        Delhi     7206704
1  1034   Pune         IND  Maharashtra     1566651

### Aliase 

In [85]:
cursor.execute('USE world')

query = """
SELECT SUM(Population) AS Pop_sum
FROM city
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\3565732755.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


Pop_sum
0  1.429560e+09

In [87]:
query = """
SELECT 
SUM(Population) AS Pop_sum,
MIN(Population) AS Pop_min,
MAX(Population) AS Pop_max,
COUNT(Population) AS Pop_count


FROM city
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\214263149.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


Pop_sum  Pop_min   Pop_max  Pop_count
0  1.429560e+09       42  10500000       4079

In [88]:
query = """
SELECT 
SUM(Population) AS Pop_sum,
MIN(Population) AS Pop_min,
MAX(Population) AS Pop_max,
COUNT(Population) AS Pop_count

FROM city

WHERE Population > 100000
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\3116099965.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


Pop_sum  Pop_min   Pop_max  Pop_count
0  1.387092e+09   100025  10500000       3558

In [89]:
query = """
SELECT 
SUM(Population) AS Pop_sum,
MIN(Population) AS Pop_min,
MAX(Population) AS Pop_max,
COUNT(Population) AS Pop_count

FROM city

WHERE Population BETWEEN 100000 AND 2000000
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_3136\2881775189.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


Pop_sum  Pop_min  Pop_max  Pop_count
0  1.007770e+09   100000  1987996       3470

Question 1: Database and Table Creation
Description: Create a database named assignment and within it, design a
table named employees with the following fields:
- employee_id (integer)
- first_name (string, max length 50)
- last_name (string, max length 50)
- salary (decimal with 2 decimal places)
- department (string, max length 50)

In [4]:
myconn = conn.connect(host = "localhost", user = "root", passwd = "12345")

cursor = myconn.cursor() # create a cursor for CRUD Operations

query = 'CREATE database assign' # query for creating a database

query = 'show databases'

cursor.execute(query)

cursor.fetchall()

query = 'USE assign'

cursor.execute(query)

query = """CREATE TABLE employees(
employee_id INT,
first_name VARCHAR(50),
last_name VARCHAR(50),
salary DECIMAL(6, 2),
department VARCHAR(50)
)"""

query = 'show tables'

tables = pd.read_sql_query(query, myconn)

print(f'Tables in the database: {tables["Tables_in_assign"]}')

NameError: name 'conn' is not defined

Question 2: Table Creation with Constraints
Description: Design a table named 'products' for storing product
information. The table should have the following fields:
- product_id (integer, primary key)
- product_name (string, max length 100)
- price (decimal with 2 decimal places, not null)
- stock_quantity (integer, not null, default 0)
- category (string, max length 50)

In [ ]:
query = """
CREATE TABLE products(
product_id INT PRIMARY KEY,
product_name VARCHAR(100),
price DECIMAL(5, 2) NOT NULL,
stock_quantity INT NOT NULL DEFAULT 0,
category VARCHAR(50)
)
"""

cursor.execute(query)

Question 3: Insert Records into 'employees' Table
Description: Given the 'employees' table with fields:
- employee_id (integer)
- first_name (string, max length 50)
- last_name (string, max length 50)
- salary (decimal with 2 decimal places)
- department (string, max length 50)
Insert 5 records into the 'employees' table with the following details:
1. Employee ID: 101, First Name: 'Ananya', Last Name: 'Kale', Salary:
60000.00, Department: 'HR'
2. Employee ID: 102, First Name: 'Pooja', Last Name: 'Patil', Salary:
75000.50, Department: 'IT'

In [147]:
myconn

In [2]:
cursor

NameError: name 'cursor' is not defined

In [1]:
query = """CREATE TABLE employees1(
employee_id INT,
first_name VARCHAR(50),
last_name VARCHAR(50),
salary DECIMAL(6, 2),
department VARCHAR(50)
)"""

cursor.execute(query)

NameError: name 'cursor' is not defined

In [162]:
query = """
INSERT INTO employees1(employee_id, first_name, last_name, salary, department)
VALUES 
    (1, 'Ananya', 'Kale', 6000.00, 'HR'),
    (2, 'Pooja', 'Patil', 7500.00, 'IT')
"""

cursor.execute(query)

In [164]:
query = """
SELECT *
FROM employees1
"""

data = pd.read_sql_query(query, myconn)

data

C:\Users\Amogh\AppData\Local\Temp\ipykernel_17488\63483172.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  data = pd.read_sql_query(query, myconn)


employee_id first_name last_name  salary department
0            1     Ananya      Kale  6000.0         HR
1            2      Pooja     Patil  7500.0         IT

In [186]:
query = 'show databases'

result = cursor.execute(query)

AttributeError: 'NoneType' object has no attribute 'fetchall'